rust/src/two_stacks_lite/mod.rs

```diff
@@ -0,0 +1,82 @@
+use std::collections::VecDeque;
+
+use alga::general::AbstractMagma;
+use alga::general::Identity;
+
+use crate::FifoWindow;
+use crate::ops::AggregateOperator;
+use crate::ops::AggregateMonoid;
+
+#[derive(Clone)]
+pub struct TwoStacksLite<BinOp>
+where
+    BinOp: AggregateMonoid<BinOp> + AggregateOperator + Clone
+{
+    queue: VecDeque<BinOp::Partial>,
+    agg_back: BinOp::Partial,
+    front_len: usize,
+}
+
+impl<BinOp> FifoWindow<BinOp> for TwoStacksLite<BinOp>
+where
+    BinOp: AggregateMonoid<BinOp> + AggregateOperator + Clone
+{
+    fn new() -> Self {
+        Self {
+            queue: VecDeque::new(),
+            agg_back: BinOp::Partial::identity(),
+            front_len: 0,
+        }
+    }
+    fn name() -> &'static str {
+        "two_stacks_lite"
+    }
+    fn push(&mut self, v: BinOp::In) {
+        let lifted = BinOp::lift(v);
+        self.queue.push_back(lifted.clone());
+        self.agg_back = self.agg_back.operate(&lifted);
+    }
+    fn pop(&mut self) -> Option<BinOp::Out> {
+        if self.queue.is_empty() {
+            None
+        } else {
+            let end = self.queue.len() - 1;
+            if self.front_len == 0 {
+                let mut i = end;
+                while i != 0 {
+                   i -= 1;
+                   self.queue[i] = self.queue[i].operate(&self.queue[i+1]);
```

Comment:
Does the underlying VecDeque support random access with an integer in constant time? If not, this line could be problematic. Could this be done with an iterator instead (or maybe two iterators, for i and i+1)?

Reply:
Yes, VecDeque supports random access in constant time. I feel like this loop should be possible using a reverse iterator, but I'm not sure if that's possible in Rust (have a reference to `i` and `i+1`). I'll look into seeing if I can change that to being iterator based.

Reply:
I guess you don't need a reference to both `i` and `i+1`. A single iterator should be enough. Something like this:

```rust
let mut iter = E - 1;
let mut accum = *iter;
iter -= 1;
loop {
    accum = (*iter).operate(accum);
    *iter = accum;
    if iter == F {
        break;
    }
    iter -= 1;
}
```

Reply:
From what I know, the `VecDeque` doesn't have a reverse iterator or a "free flowing" iterator, so we might need to work around this. Also, the backing storage of `VecDeque` is a ring buffer. In order for us to have a version of Two-Stacks Lite that can use both a ring-buffer-based backing queue and a chunked-array backing queue, would it make sense to work out a new trait and/or extend the `VecDeque` with what we will eventually need for a queue backed by a chunked-array list and later DABA?

Reply:
@scotts Great. I missed the fact that `VecDeque` implements `DoubleEndedIterator`, which should be sufficient for our use.

Reply:
I think there should be one or more chunked array queue implementations floating around on https://crates.io/. The solution I had in mind before was to use:
* a `std::collections::LinkedList` for connecting the chunks
* a `std::vec::Vec` (or array `[T;N]`) for storing each chunk.
* a `std::collections::linked_list::CursorMut` for traversing between chunks.
* a higher-level Cursor, building on top of `CursorMut` for traversing between elements of chunks.

A cursor is like an iterator but does not consume the thing it is iterating over. With `DoubleEndedIterator`, it might be a problem because you cannot iterate over the same element twice:

```rust
let numbers = vec![1, 2, 3, 4, 5, 6];

let mut iter = numbers.iter();

assert_eq!(Some(&1), iter.next());
assert_eq!(Some(&6), iter.next_back());
assert_eq!(Some(&5), iter.next_back());
assert_eq!(Some(&2), iter.next());
assert_eq!(Some(&3), iter.next());
assert_eq!(Some(&4), iter.next());
assert_eq!(None, iter.next());
assert_eq!(None, iter.next_back());
```

If you need multiple cursors, you might have to create them using an unsafe pointer. I think it is not possible to use a `std::cell::RefCell` + `std::rc::Rc` because the cursors must mutably borrow the linked list which they are traversing for their whole lifetime. With `RefCell`, Rust would panic if two cursors exist simultaneously. There should hopefully be no problem as long as you ensure that the cursors do not point to something which is deallocated.

```rust
fn main() {
    use std::collections::LinkedList;
    let mut l: LinkedList<[i32; 5]> = LinkedList::new();

    let (mut c0, mut c1, mut c2) = unsafe {
        let l = &mut l;
        let l = l as *mut LinkedList<[i32; 5]>;
        let c0 = l.as_mut().unwrap().cursor_back_mut();
        let c1 = l.as_mut().unwrap().cursor_back_mut();
        let c2 = l.as_mut().unwrap().cursor_back_mut();
        (c0, c1, c2)
    };

    l.push_back([0,1,2,3,4]);
    l.push_back([5,6,7,8,9]);

    c0.move_next();
    c1.move_next();
    c2.move_next();
    c2.move_next();

    l.push_front([4,3,2,1,0]);

    c0.move_prev();

    println!("{:?}", l);

    assert_eq!(c0.current(), Some(&mut [4,3,2,1,0]));
    assert_eq!(c1.current(), Some(&mut [0,1,2,3,4]));
    assert_eq!(c2.current(), Some(&mut [5,6,7,8,9]));
}
```

Reply:
I asked in the Rust community and was recommended this crate: https://crates.io/crates/bucket_queue

Reply:
@segeljakt, your comment about implementation strategies for a Rust version of the chunked-array queue is so helpful that I'm going to create a new issue around it. Myself, Martin and Kanat have talked about what it could potentially look like, so that issue will become the center of that discussion.

Reply:
Changed to use iterators.